### PR TITLE
[core] Add cache for snapshots in CachingCatalog

### PR DESCRIPTION
**Pull request by JingsongLi**
### Purpose

Considering that all snapshot files are immutable, we are considering adding cache to them. It is safe to cache them in Table Object.

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/tag/Tag.java [hunk 1]
```diff
@@ -201,4 +179,28 @@ public boolean equals(Object o) {
...
+
+    public static Tag fromPath(FileIO fileIO, Path path) {
+        try {
+            return tryFromPath(fileIO, path);
+        } catch (FileNotFoundException e) {
+            throw new RuntimeException(e.getMessage(), e);
+        }
+    }
+
+    public static Tag tryFromPath(FileIO fileIO, Path path) throws FileNotFoundException {
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/tag/Tag.java:197]: Why not cache Tag and Changelog like Snapshot？

File: paimon-core/src/main/java/org/apache/paimon/Snapshot.java [hunk 2]
```diff
@@ -437,4 +413,41 @@ public enum CommitKind {
...
+
+    public static Snapshot tryFromPath(FileIO fileIO, Path path) throws FileNotFoundException {
+        try {
+            Snapshot snapshot = SNAPSHOT_CACHE.getIfPresent(path);
+            if (snapshot == null) {
+                snapshot = Snapshot.fromJson(fileIO.readFileUtf8(path));
+                SNAPSHOT_CACHE.put(path, snapshot);
+            }
+            return snapshot;
+        } catch (FileNotFoundException e) {
```
[the remark this answers is not in the document]
Comment (JingsongLi) [reply paimon-core/src/main/java/org/apache/paimon/Snapshot.java:440]: There is no other IOException.

File: paimon-core/src/main/java/org/apache/paimon/catalog/CachingCatalog.java [hunk 3]
```diff
@@ -316,4 +319,36 @@ public void refreshPartitions(Identifier identifier) throws TableNotExistExcepti
...
+                    .executor(Runnable::run)
+                    .build();
+
+    public static void invalidateMetaCacheForPrefix(Path tablePath) {
+        String path = tablePath.toString();
+        invalidateMetaCacheForPrefix(SNAPSHOT_CACHE, path);
+        invalidateMetaCacheForPrefix(SCHEMA_CACHE, path);
+    }
+
+    private static void invalidateMetaCacheForPrefix(Cache<Path, ?> cache, String tablePath) {
```
[the remark this answers is not in the document]
Comment (JingsongLi) [reply paimon-core/src/main/java/org/apache/paimon/catalog/CachingCatalog.java]: deleted

File: paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/HiveCatalog.java [hunk 4]
```diff
@@ -680,6 +681,9 @@ protected void dropTableImpl(Identifier identifier) {
                                     false,
                                     true));
 
+            Path path = getTableLocation(identifier);
+            invalidateMetaCacheForPrefix(path);
```
[the remark this answers is not in the document]
Comment (JingsongLi) [reply paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/HiveCatalog.java]: deleted

File: paimon-core/src/main/java/org/apache/paimon/tag/Tag.java (hunk 1 above)
Comment (JingsongLi) [reply paimon-core/src/main/java/org/apache/paimon/tag/Tag.java:197]: Tag is mutable, it is danger.

**Review by yuzelin — APPROVED** [review]
+1

**Comment by ranxianglei** [comment]
mark!

**Pull request merged**